theano/d3viz/d3viz.py

```diff
@@ -0,0 +1,99 @@
+"""Dynamic visualization of Theano graphs.
+
+Author: Christof Angermueller <cangermueller@gmail.com>
+"""
+
+import os
+import os.path as pt
+import shutil
+import re
+
+from formatting import PyDotFormatter
+
+__path__ = pt.dirname(pt.realpath(__file__))
+
+
+def replace_patterns(x, replace):
+    """Replace patterns `replace` in x."""
+    for from_, to in replace.items():
+        x = x.replace(str(from_), str(to))
+    return x
+
+
+def escape_quotes(s):
+    """Escape quotes in string."""
+    s = re.sub(r'''(['"])''', r'\\\1', s)
+    return s
+
+
+def d3viz(fct, outfile, copy_deps=True, *args, **kwargs):
+    """Create HTML file with dynamic visualizing of a Theano function graph.
+
+    :param fct: A compiled Theano function, variable, apply or a list of
+                variables.
+    :param outfile: The output HTML file.
+    :param copy_deps: Copy javascript and CSS dependencies to output directory.
+    :param *args, **kwargs: Arguments passed to PyDotFormatter.
+
+    In the HTML file, the whole graph or single nodes can be moved by drag and
+    drop. Zooming is possible via the mouse wheel. Detailed information about
+    nodes and edges are displayed via mouse-over events. Node labels can be
+    edited by selecting Edit from the context menu.
+
+    Input nodes are colored in green, output nodes in blue. Apply nodes are
+    ellipses, and colored depending on the type of operation they perform. Red
+    ellipses are transfers from/to the GPU (ops with names GpuFromHost,
+    HostFromGpu).
+
+    Edges are black by default. If a node returns a view of an
+    input, the input edge will be blue. If it returns a destroyed input, the
+    edge will be red.
```

Comment:
Docstring should be numpydoc style.  See https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt